revscoring/utilities/fetch_idioms.py

```diff
@@ -63,48 +64,46 @@ def is_idiom(phrase):
     return True
 
 
-def combine_words(first_word, following_words):
-    """
-        Combines the first word and following words into one regex
-        Sample:
-        first_word (word1|word2|...|wordx)
-    """
-    if len(following_words) == 1:
-        idiom_regex = '{} {}'.format(
-            first_word, "".join(following_words))
+def construct_trie(word_list):
+    root = OrderedDict()
+    for word in word_list:
+        level = root
+        # Start at root node
+        for letter in word:
+            if letter not in level:
+                # Create a child node
+                level[letter] = OrderedDict()
+            # Set new level to the child node
+            level = level[letter]
+    return root
+
+
+def convert_trie_to_regex(trie):
+    # Extract the nodes and the sub-trees
+    keys = trie.keys()
+    values = trie.values()
+
+    # Return an empty string if node is a leaf node
+    if len(values) == 0:
+        return keys if len(keys) else ''
     else:
-        idiom_regex = '{} ({})'.format(
-            first_word, "|".join(following_words))
-    if not idiom_regex.isspace():
-        return idiom_regex
+        combo = []
+        # Traverse depth first
+        for key, value in zip(keys, values):
+            combo.append(key + convert_trie_to_regex(value))
+        if len(combo) > 1:
+            return "(?:{})".format('|'.join(combo))
+        else:
+            return ''.join(combo)

```

Comment:
Shouldn't we apply `re.escape(...)` somwhere?

Reply:
@he7d3r My first approach to this was to generate a trie out of the regular list, then convert it into optimized regex instead of piping the words together. Hence, with that approach, we can't allow any regex in the list of words.

Reply:
It doesn't seem like we need this code anymore, do we?

Reply:
We don't.